### ISSUE TITLE
Handle the deletion of a managedcluster resource

### ISSUE THREAD

**Issue by maleck13** [Kuadrant/multicluster-gateway-controller#303]
**What**

When a managedcluster is deleted, we need to ensure that the records for any gateway on that cluster are cleaned up. 
Currently the records are not cleaned up and the controller has several reconcile errors

If the managedcluster is deleted:
- grace period should be ignored
- dns records should be tidied up

It should also be possible to modify things like the DNSPolicy and GEO attributes of existing clusters.
Currently the DNSPolicy controller wont apply a new country code as it is failing to find a cluster

```
ERROR	Reconciler error	{"controller": "dnspolicy", "controllerGroup": "kuadrant.io", "controllerKind": "DNSPolicy", "DNSPolicy": {"name":"prod-web","namespace":"multi-cluster-gateways"}, "namespace": "multi-cluster-gateways", "name": "prod-web", "reconcileID": "97742d88-5b84-44c7-92b0-b96ba4e0985e", "error": "get cluster gateway failed: ManagedCluster.cluster.open-cluster-management.io \"kind-mgc-workload-1\" not found"}

```

**Comment by david-martin** [comment]
Captured various state after the delete of the workload1 ManagedCluster post ocm walkthrough.
https://gist.github.com/david-martin/49421cde1fd12ac4d94fd3a9cfade228

**Comment by david-martin** [comment]
After deleting 1 out of 2 ManagedClusters, the state of resources is as follows:

* The PlacementDecision resource is updated (by OCM) to have only 1 cluster in the `status.decisions`. The Placement resource is also updated (by OCM) to set the PlacementSatisfied status condition to `False` and a reason of `NotAllDecisionsScheduled`
* The Gateway resource still exists in both clusters. This happens presumably because the mgc controller does not have logic to delete a Gateway (or there's a bug) when the placement decisions change. So, the ManifestWork that contains the Gateway continues to exist.
* The DNSRecord resource in the hub still has both endpoints from both Gateways, including all dnsNames (A & CNAME records). 
* The DNSPolicy resource in the hub has a `Ready` status condition of `False`, with a reason of `ReconciliationError` and message of `get cluster gateway failed: ManagedCluster.cluster.open-cluster-management.io "kind-mgc-workload-1" not found`

Based on the above, here's what I'd recommend needs to be done:

* The PlacementDecision and Placement resources seem to be behaving as expected. No changes needed there.
* The ManifestWork for the Gateway placed in the ManagedCluster that was deleted should also be deleted. This will cause OCM to delete the Gateway resource in that cluster. This should in turn trigger a reconcile of the Gateway in the hub, updating the list of `addresses` and `listeners`. This will need to be verified.
* With the changes above, it should trigger the dns controller & update the DNSRecord resource to remove the endpoints that correspond to the deleted cluster. A & CNAME records in Route53 should also be removed. This will need to be verified.
* Also with the changes above, the dns policy controller should no longer look for the deleted ManagedCluster and should no longer result in a `ReconciliationError`. This will need to be verified.

@maleck13 I think I've captured all the touch points here.
To summarise, the only change should be to remove the Gateway (via the ManifestWork) from the ManagedCluster that was deleted. After that, everything else _should_ fall into place, but needs to be verified.

**Comment by mikenairn** [comment]
> * The ManifestWork for the Gateway placed in the ManagedCluster that was deleted should also be deleted. This will cause OCM to delete the Gateway resource in that cluster. This should in turn trigger a reconcile of the Gateway in the hub, updating the list of `addresses` and `listeners`. This will need to be verified.

I think if this is done, and the placed clusters no longer includes the deleted one, it should be enough. As mentioned on chat, we will need to make sure this doesn't cause a reconcile of the dnspolicy on delete events for the managed cluster, or it might process the policy before the gateway controller has updated things. https://github.com/Kuadrant/multicluster-gateway-controller/commit/45d434f86ed0bc5641f2d1791b5c078368be1145#diff-162ceed42032f121fd671b5884a611158c00f31e292b7602e48d5671ca7fe01fR243-R246
